compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -12810,6 +12810,149 @@ TR::Node *su2dSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
+
+      // Get and check the integer/long constant values
+      if (isInt)
+         {
+         oldConst1 = opNode->getSecondChild()->getConstValue();
+         oldConst2 = secondChild->getConstValue();
+         newConst = isAddOp ? (oldConst2 - oldConst1) : (oldConst2 + oldConst1);
+
+         // Get type ranges to check for overflow or underflow
+         int64_t signedMax = 0, signedMin = 0, unsignedMax = 0, unsignedMin = 0;
+         if (opNode->getOpCode().is1Byte())
+            {
+            signedMax   = TR::getMaxSigned<TR::Int8>();
+            signedMin   = TR::getMinSigned<TR::Int8>();
+            unsignedMax = TR::getMaxUnsigned<TR::Int8>();
+            unsignedMin = TR::getMinUnsigned<TR::Int8>();
+            }
+         else if (opNode->getOpCode().is2Byte())
+            {
+            signedMax   = TR::getMaxSigned<TR::Int16>();
+            signedMin   = TR::getMinSigned<TR::Int16>();
+            unsignedMax = TR::getMaxUnsigned<TR::Int16>();
+            unsignedMin = TR::getMinUnsigned<TR::Int16>();
+            }
+         else if (opNode->getOpCode().is4Byte())
+            {
+            signedMax   = TR::getMaxSigned<TR::Int32>();
+            signedMin   = TR::getMinSigned<TR::Int32>();
+            unsignedMax = TR::getMaxUnsigned<TR::Int32>();
+            unsignedMin = TR::getMinUnsigned<TR::Int32>();
+            }
+         else
+            return node;
+
+         if ((!isUnsignedCompare && (newConst > signedMax || newConst < signedMin)) ||
+                 (isUnsignedCompare && (newConst > unsignedMax || newConst < unsignedMin)))
+            return node;
+         }
+      else  // Long
+         {
+         if (isUnsignedCompare)
+            {
+            uint64_t oldUConst1 = opNode->getSecondChild()->getUnsignedLongInt();
+            uint64_t oldUConst2 = secondChild->getUnsignedLongInt();
+
+            if (isAddOp && (oldUConst2 > oldUConst1))
+               newUConst = oldUConst2 - oldUConst1;
+            else if (isSubOp && (oldUConst2 <= TR::getMaxUnsigned<TR::Int64>() - oldUConst1))
+               newUConst = oldUConst2 + oldUConst1;
+            else
+               return node;
+            }
+         else // signed long compares
+            {
+            oldConst1 = opNode->getSecondChild()->getConstValue();
+            oldConst2 = secondChild->getConstValue();
+
+            if (isAddOp &&
+                    !(oldConst1 > 0 && (oldConst2 < TR::getMinSigned<TR::Int64>() + oldConst1)) &&
+                    !(oldConst1 < 0 && (oldConst2 > TR::getMaxSigned<TR::Int64>() + oldConst1)))
+               newConst = oldConst2 - oldConst1;
+            else if (isSubOp &&
+                     !(oldConst1 > 0 && (oldConst2 > TR::getMaxSigned<TR::Int64>() - oldConst1)) &&
+                     !(oldConst1 < 0 && (oldConst2 < TR::getMinSigned<TR::Int64>() - oldConst1)))
+               newConst = oldConst2 + oldConst1;
+            else
+               return node;
```

Comment:
The overflow checking logic here looks good! 👍 

Is it possible to abstract this a bit so that it can be used for the narrower cases as well? Widening to prevent overflow is good, but since this code is still needed anyway, maybe it can be used to simplify the function